File: javaprojects/common/src/main/java/minigames/rendering/RenderingPackage.java
package minigames.rendering;

import java.util.List;
import io.vertx.core.json.JsonObject;

/**
 * Returned by a game to update the UI state of the client.
 * 
 * Rendering commands are just given as a list of JSON objects. 
 * This lets GameClients and GameServers implement bespoke rendering commands at will.
 * 
 * There are a small set of rendering commands known natively to the client
 */
public record RenderingPackage(
    GameMetadata metadata,
    List<JsonObject> renderingCommands
) {

  public static RenderingPackage fromJson(JsonObject json) {
    return new RenderingPackage(
        json.getJsonObject("metadata").mapTo(GameMetadata.class),
        json.getJsonArray("renderingCommands").stream().map((o) -> (JsonObject)o).toList()
    );
  }

}

File: scalaprojects/scalafx-client/src/main/scala/minigames/rendering/RenderingPackage.scala
package minigames.rendering

import minigames.rendering.GameMetadata
import scala.util.*
import com.fasterxml.jackson.core.`type`.TypeReference
import com.fasterxml.jackson.databind.JsonNode
import io.vertx.core.json.*
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.deser.std.StdDeserializer
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.node.ObjectNode
import minigames.commands.mapper

import scala.jdk.CollectionConverters.*

case class RenderingPackage(metadata:GameMetadata, renderingCommands:Seq[JsonObject])

object RenderingPackage:
    def fromJson(json:JsonObject) = 
        RenderingPackage(
            GameMetadata.fromJson(json.getJsonObject("metadata")),
            (for 
                case o:JsonObject <- json.getJsonArray("renderingCommands").iterator().asScala
            yield o).toSeq
        );
